### PR TITLE
Feature/reduce websocket subscriptions

### PR DESCRIPTION
**Pull request by sleiss**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I added multiple integration tests (Jest) related to the features
- [x] Client: I documented the TypeScript code using JSDoc style.

### Motivation and Context
Currently, a new WebSocket subscription is opened for every unrated exercise where the due date passed and for every started programming exercise for every student in the course overview. Also, for TAs and higher, a WebSocket subscription is opened for every student once they view an exercise overview.

In large courses like EIST, around 1500 students use the system at the same time. Every one of these students typically has quite some unrated exercises and started programming exercises, sometimes leading to > 10 subscriptions per students. The subscribe-mechanism is quite exhausting for the server which already lead to problems during one quiz in a live lecture as there was a large queue with unprocessed subscriptions, decreasing the user experience and functionality as other WebSocket messages could also not be processed.

This PR creates one subscription for all `/newResults`-messages and one for every `/newSubmission`-messages for each student. Also, TAs and higher will now open one subscription per exercise, instead of one per participation.

### Description
A new topic `/user/topic/newResults` is introduced which every client can independently subscribe to. The server will then send the new results to this topic for the student(s) owning the participation instead of a topic only used by this participation.
The new result will also be broadcasted to the topic `/topic/exercise/{exerciseId}/newResults`, where TAs and higher can subscribe to.
A security check was added in the server that ensures that only TAs and higher can subscribe to the `/topic/exercise/{exerciseId}/newResults` topic.

In the client, at several occurrences, it has to be specified if a subscription to a new result for a participation is for a student or for an instructor. This is done by adding a property `personalParticipation` to several components. This value will be set be their parent components, depending on the context the child component is used (e.g. in the course overview for a student, the `UpdatingResultComponent` will be used in a personal context whereas it will be used in a non-personal context in the exercise view for instructors (where the state of the template & solution repositories are shown)).

The client also keeps a map which stores which topic was used by which participation, therefore ensuring that every topic will only be subscribed once.
This logic also ensures that an unsubscription from one specific WebSocket topic will only occur if no other participations rely on this topic.

#### The difference between the previous subscriptions and the current subscriptions can be found in the following tables:
| Old subscriptions            | Student                                              | TA/Instructor/Admin                                  |
|------------------------------|------------------------------------------------------|------------------------------------------------------|
| Inform about new results     | /topic/participation/{participationId}/newResults    | /topic/participation/{participationId}/newResults    |
| Inform about new submissions | /topic/participation/ {particpationId}/newSubmission | /topic/participation/ {particpationId}/newSubmission |

| New subscriptions            | Student                   | TA/Instructor/Admin                        |
|------------------------------|---------------------------|--------------------------------------------|
| Inform about new results     | /user/topic/newResults    | /topic/exercise/{exerciseId}/newResults    |
| Inform about new submissions | /user/topic/newSubmission | /topic/exercise/{exerciseId}/newSubmission |

The number websocket subscriptions is reduced as 2 subscriptions per student are enough and also the instructors only need 2 subscriptions per exercise instead of per participation.

### Steps for Testing 
In general, results and subscriptions should be updated without reload wherever they are updated before this PR, but in a more performant way.

1. Log in to Artemis
2. Navigate to Course Administration
3. Create a text exercise
4. Participate as a student
5. Assess the submission of the student
6. The student should be informed without reload about the new result

Peek 2020-06-04 11-46

1. Log in to Artemis
2. Navigate to Course Administration
3. Create a programming exercise
4. Click on the View-button for the exercise.
5. Scroll down, the status of the template and solution participation should be displayed.
6. Re-trigger the builds for template and solution participations. Only the participation that you triggered should be changed to "Building and testing..." and the received result should be displayed. (**NOTE**: this can take some time as the build is processed in the background on the CI-server, but it should normally take not significantly longer than it took for me locally (see GIF))
Peek 2020-06-04 12-01

7. Start the programming exercise as a student
8. Change some code and submit, the text should change to "Testing and Building..."
9. After the build was executed, the result should be displayed.

Peek 2020-06-04 15-10

### CODE REVIEW

**Comment by tiffca** [comment]
I tested this on test server 1.
The first part worked fine. Current date has to be after assignment due date and assessment due date, then the student is informed without reload about the new result, as expected. 

The second part didn't work for me on test server 1 (deployed [this build](https://bamboo.ase.in.tum.de/browse/ARTEMIS-WEBAPP1150-33))
Steps
1. Created a new programming exercise, used default options, selected allow code editor. 
2. Went to view -> reload  template/solution participation - this worked the first time (displays building and shows the results). It took about 15 seconds, and 8 seconds. Works as expected
3. As a student, started the programming exercise 
4. As an instructor, if I go back and try again to view -> reload template/solution participation - this option no longer seems to work, the button doesn't display building or the results, the button doesn't seem to work. Expectation: the button should work the same as before
5. As a student, open the modeling editor, make a change, and submit. I don't see the status change to building. Expectation: the status should display building 
6. After some time, about 10 seconds, the status does display (e.g. "build failed 0/5 tests passed"). 
7. As a student, make additional changes in the code editor. I'm not able to submit, I can only save. Expected to be able to submit again.

**Comment by Alexanderkrp** [comment]
Tested both scenarios and everything worked like expected 👍

File: src/main/java/de/tum/in/www1/artemis/config/Constants.java [hunk 1]
```diff
@@ -32,7 +32,11 @@
 
     public static final String FILEPATH_ID_PLACHEOLDER = "PLACEHOLDER_FOR_ID";
 
-    public static final String PARTICIPATION_TOPIC_ROOT = "/topic/participation/";
+    public static final String EXERCISE_TOPIC_ROOT = "/topic/exercise/";
+
+    public static final String NEW_SUBMISSION_TOPIC = "/topic/newSubmission";
+
+    public static final String NEW_RESULT_TOPIC = "/topic/newResults";
```
Comment (madwau) [inline src/main/java/de/tum/in/www1/artemis/config/Constants.java:39]: Would be nice if singular and plural are consistent for `/topic/newSubmission` and `/topic/newResults`.

File: src/main/java/de/tum/in/www1/artemis/config/websocket/WebsocketConfiguration.java [hunk 2]
```diff
@@ -228,4 +249,10 @@ private boolean isParticipationOwnedByUser(Principal principal, Long participati
         StudentParticipation participation = participationService.findOneStudentParticipation(participationId);
         return participation.isOwnedBy(principal.getName());
     }
+
+    private boolean isUserTAOrHigherForExercise(Principal principal, Long exerciseId) {
+        Exercise exercise = exerciseService.findOne(exerciseId);
+        User user = userService.getUserWithGroupsAndAuthorities(principal.getName());
+        return authorizationCheckService.isAtLeastInstructorForExercise(exercise, user);
```
Comment (madwau) [inline src/main/java/de/tum/in/www1/artemis/config/websocket/WebsocketConfiguration.java:256]: The method name suggests that it is checked whether the user is a TA or above but the actual check is done for instructor or above.

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java [hunk 3]
```diff
@@ -529,13 +528,27 @@ private String getProgrammingExerciseAllExerciseBuildsTriggeredTopic(Long progra
...
     public void notifyUserAboutSubmission(ProgrammingSubmission submission) {
-        String topic = Constants.PARTICIPATION_TOPIC_ROOT + submission.getParticipation().getId() + Constants.PROGRAMMING_SUBMISSION_TOPIC;
-        messagingTemplate.convertAndSend(topic, submission);
+        if (submission.getParticipation() instanceof StudentParticipation) {
+            StudentParticipation studentParticipation = (StudentParticipation) submission.getParticipation();
+            studentParticipation.getStudents().forEach(user -> messagingTemplate.convertAndSendToUser(user.getLogin(), NEW_SUBMISSION_TOPIC, submission));
+        }
+
+        if (submission.getParticipation() != null && submission.getParticipation().getExercise() != null) {
+            String exerciseTopic = EXERCISE_TOPIC_ROOT + submission.getParticipation().getExercise().getId() + "/newSubmission";
```
Comment (madwau) [inline src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java:537]: Maybe there can be a static builder method that constructs this topic? At least `/newSubmission` should not be hardcoded and appear in multiple places.

File: src/main/java/de/tum/in/www1/artemis/service/WebsocketMessagingService.java [hunk 4]
```diff
@@ -40,9 +48,46 @@ public void broadcastNewResult(Participation participation, Result result) {
...
+
+        // Send to tutors, instructors and admins
+        messagingTemplate.convertAndSend(getResultDestination(participation.getExercise().getId()), result);
 
         // recover the participation because we might want to use it again after this method
         result.setParticipation(originalParticipation);
     }
+
+    /**
+     * Returns true if the given destination should be handled by this service
```
Comment (madwau) [inline src/main/java/de/tum/in/www1/artemis/service/WebsocketMessagingService.java:65]: Does not really explain what the method is actually checking.

File: src/main/webapp/app/exercises/programming/participate/programming-submission.service.ts [hunk 5]
```diff
@@ -52,11 +52,15 @@ export class ProgrammingSubmissionService implements IProgrammingSubmissionServi
...
     private DEFAULT_EXPECTED_RESULT_ETA = 2 * 60 * 1000;
-    private SUBMISSION_TEMPLATE_TOPIC = '/topic/participation/%participationId%/newSubmission';
+    private SUBMISSION_TEMPLATE_TOPIC = '/topic/exercise/%exerciseId%/newSubmission';
 
     private resultSubscriptions: { [participationId: number]: Subscription } = {};
     // participationId -> topic
     private submissionTopicsSubscribed = new Map<number, string>();
+
+    // participationId -> exerciseId
+    private participationToExercise = new Map<number, number>();
```
Comment (madwau) [inline src/main/webapp/app/exercises/programming/participate/programming-submission.service.ts:62]:
```suggestion
    private participationIdToExerciseId = new Map<number, number>();
```

File: src/main/java/de/tum/in/www1/artemis/config/websocket/WebsocketConfiguration.java (hunk 2 above)
Comment (sleiss) [reply src/main/java/de/tum/in/www1/artemis/config/websocket/WebsocketConfiguration.java:256]: Very good catch! I'll fix it.

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java (hunk 3 above)
Comment (sleiss) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java:537]: I've created a builder for it.

File: src/main/java/de/tum/in/www1/artemis/config/Constants.java (hunk 1 above)
Comment (sleiss) [reply src/main/java/de/tum/in/www1/artemis/config/Constants.java:39]: I've changed it to `/topic/newSubmissions` (and removed some more duplications).

File: src/main/java/de/tum/in/www1/artemis/service/WebsocketMessagingService.java (hunk 4 above)
Comment (sleiss) [reply src/main/java/de/tum/in/www1/artemis/service/WebsocketMessagingService.java:65]: You are right, I've adjusted the documentation.

Comment (sleiss) [reply src/main/java/de/tum/in/www1/artemis/service/WebsocketMessagingService.java:65]: @madwau 
> Left some code comments.

Thanks for the feedback, I addressed the comments.

Comment (madwau) [reply src/main/java/de/tum/in/www1/artemis/service/WebsocketMessagingService.java]: Great!

**Pull request merged**